[source-verification 2] Reimplement verify-source (#27213)
MystenLabs/sui: docs/content/develop/manage-packages/source-verification.mdx

```bash
sui client verify-source <PACKAGE_PATH> <PACKAGE_ID>
```

- `PACKAGE_PATH` is the path to the source package (the directory containing `Move.toml`).
- `PACKAGE_ID` is the address of the on-chain package to compare against.

The command builds against whichever network your CLI is pointed at, so switch to the right
environment first (for example, `sui client switch --env mainnet`).

On success it prints `Source verification succeeded!`. Otherwise it lists every difference it found.

## What is verified

Only the **root** package is verified: the bytecode of its modules and its linkage table (the
versions of its dependencies). The source of the dependencies themselves is not checked. To verify a
dependency, run `verify-source` again against that dependency's own source and address.

The command performs no compilation itself. It resolves the toolchain version the package was
published with, downloads that release's `sui` binary if it is not already cached under
`$MOVE_HOME/binaries`, and uses it to build the source. Because the on-chain bytecode was produced by
that same toolchain, a matching build is strong evidence that the source is authentic.

> :::note

Author: should maybe be stronger and/or mentioned in the first paragraph.

Author: In fact, I wonder how hard it would be to detect unused macros and produce a warning. Probably requires getting into bed with the compiler, in which case it's probably not worth the complexity.

Author: [CLAUDE] Strengthened the caveat: it now spells out that the bodies of macros that *are* called are inlined and therefore covered, while only comments, formatting, and uncalled-macro bodies can differ. Kept it as a prominent `:::note` right after "What is verified" rather than in the first paragraph. See source-verification.mdx.

Verification compares **bytecode**, not source text. Changes that do not affect the compiled output —
comments, formatting, and the bodies of `macro` functions that are never called — can differ between
the source and what was published and still verify successfully. Conversely, a package that verifies
has bytecode identical to what is on chain.

:::

> ## Choosing the toolchain version

Author: Maybe we should move this section into the "when verification fails" below?

Author: [CLAUDE] Done — it is now the "The toolchain version is not recorded" subsection under "When verification fails".

`Move.lock` for packages published before v1.63). Some publications do not record it — notably
packages published before v1.23, and some multisig or offline publishing flows. In that case the
command stops and asks you to supply the version:

```bash
sui client verify-source <PACKAGE_PATH> <PACKAGE_ID> --toolchain-version 1.71.1
```

If you do not know the exact version, the release that was current when the package was published is
a good first guess. Compiler output is stable across neighbouring releases, so an adjacent version
usually reproduces the same bytecode. A version from far in the past or future may not, because the
bytecode file format changes occasionally.

`--toolchain-version` is also the workaround when a package's recorded toolchain cannot be used at
all — for example, a release whose framework revision is no longer available. The command fails early
in that case and suggests trying a different version.

## When verification fails

A failure does not necessarily mean the source is wrong. Reproducing a build is exacting, and there
are a few common reasons a genuine source fails to verify.

> 

Author: We might recommend publishing a tag with a fixed source package once you get verification to pass (including the toolchain-version and pinned deps).

Author: [CLAUDE] Added a "Making a package verifiable" section recommending tagging the exact published commit, plus recording the toolchain-version and pinning deps to hashes. source-verification.mdx.

### You are on the wrong commit

The most common cause is source drift: the working tree is close to, but not exactly, what was
published. The command reports which modules differ, and a small number of named modules is a strong
hint that only those files have changed since publication.

Check out the exact commit that was published. The commit that *records* a package's `published-at`
in source control is often a later commit than the one that was actually built, so search for when
the address first appeared:

```bash

> git log -S <PACKAGE_ID> -- '**/Published.toml' '**/Move.lock'

Author: Is there a simple CLI command that can tell you the date that something was published, and then we could suggest using git log with that date? I wouldn't want a hugely complicated snippet here, but I don't think the way it's described here is correct.

Author: [CLAUDE] Reworked so it no longer needs a publish-date lookup: for the common case (a framework dependency) it points at `crates/sui-framework-snapshot/manifest.json` (revision per protocol version) or the `sui` release current at publish time. See the "Dependencies are not pinned to commit hashes" section.

published with, downloads that release's `sui` binary if it is not already cached under
`$MOVE_HOME/binaries`, and uses it to build the source. Because the on-chain bytecode was produced by
that same toolchain, a matching build is strong evidence that the source is authentic.

:::note

Verification compares **bytecode**, not source text. Anything that does not affect the compiled
output — comments, formatting, and the bodies of `macro` functions that are never called — can differ
between the source and what was published and still verify successfully. (The bodies of macros that
*are* called are inlined into their callers, so they do affect the bytecode and are covered.)
Conversely, a package that verifies has bytecode identical to what is on chain.

:::

## Making a package verifiable

If you publish a package and want others to be able to verify it out of the box:

- **Pin every dependency to a commit hash**, not a branch or tag. The current package system does
  this automatically; see [Dependencies are not pinned to commit
  hashes](#dependencies-are-not-pinned-to-commit-hashes) below for older packages.

> - **Tag the exact commit that was published**, so verifiers can check out the precise source. The

Author: This is a little misleading - you want to tag the version that verifies. For example, if you added a Published.toml file you'd want to tag that commit.

Author: What we should say is something about finding the actual source code that you published

Author: [CLAUDE] Reworded: it now says to tag a commit whose source *verifies* — not necessarily the original publish commit; if you added a `Published.toml` or pinned deps to make it verify, tag that commit. source-verification.mdx.

  commit that records the `published-at` address in source control is often a later commit than the
  one that was actually built.
- **Make sure the publication records the toolchain version.** Modern publishing writes it to
  `Published.toml`. If yours does not (see below), you can add it.

> If a package's publication does not record its address or toolchain — for example, it was published

Author: We should either link to the appropriate section of the docs if there's already a page describing the format, or else give a little example here

Author: [CLAUDE] Added an inline `Published.toml` example (one `[published.<env>]` section with `chain-id`, `published-at`, `original-id`, `version`, `toolchain-version`). source-verification.mdx.

hint that only those files have changed since publication.

Check out the exact commit that was published. Because the commit that *records* a package's
`published-at` in source control is often later than the one that was built, search for when the
address first appeared and check out the earliest such commit:

```bash
git log --all -S <PACKAGE_ID> -- '**/Published.toml' '**/Move.lock'
```

### The toolchain version is not recorded

By default the toolchain version comes from the package's publish metadata. Some publications do not
record it — notably packages published before v1.23, and some multisig or offline flows. In that case
the command stops and asks you to supply the version:

```bash
sui client verify-source <PACKAGE_PATH> --toolchain-version 1.71.1
```

If you do not know the exact version, the release that was current when the package was published is a
good first guess. Compiler output is stable across neighbouring releases, so an adjacent version
usually reproduces the same bytecode; a version from far in the past or future may not, because the
bytecode file format changes occasionally.

> `--toolchain-version` is also the workaround when a package's recorded toolchain cannot be used at

Author: I think the list of supported versions should either be here or possibly in the error message for unsupported versions.

Author: [CLAUDE] Added a supported-versions note to the docs: any release with a binary for your platform (roughly v1.8+) can be used as the rebuild toolchain, and packages published from v1.23 verify from metadata without extra arguments. The full sweep result (coverable range + the three gaps: no binary <= v1.8.1, protocol-config deserialization v1.25.1-v1.29.2, v1.64.1's missing framework rev) is now documented in DESIGN.md's testing section. I can also surface the specifics in the download error message if you'd prefer that over the docs.

Author: [CLAUDE] Now surfaced in the error message too, hardcoded (no git/network query): a toolchain download or build failure appends a targeted suggestion. The versions known not to work for verification — v1.8.1 and earlier (no platform binary) and v1.64.1 (gc-ed framework) — get a hint naming a nearby working release (e.g. "1.63.4 or 1.65.2" for 1.64.1); anything else gets generic "try an adjacent release" advice. See `toolchain_suggestion` in error.rs (with a test).

Author: [CLAUDE] Restructured per your suggestion: rather than appending on failure, `check_toolchain_version` now runs up front and fails fast with a precise message for the known gaps — e.g. "The recorded toolchain sui 1.64.1 cannot verify this package: v1.64.x pins a framework revision that is no longer available. Passing --toolchain-version 1.65.2 may work." (<= v1.8.1 -> 1.9.0; v1.64.x -> 1.65.2). See `check_toolchain_version` in lib.rs, with a test.

all — for example, a release whose framework revision is no longer available. The command fails early
in that case and suggests trying a different version.

### Dependencies are not pinned to commit hashes

The current package system pins every dependency to a specific commit hash, so its builds are
reproducible. Packages published under the older system (before v1.63) may record a dependency's
revision as a **branch** rather than a hash. Rebuilding then fetches whatever that branch points at
now, which is almost never what was published — and often does not even compile against the older
toolchain. When this happens, the command reports each dependency that is not pinned to a hash.

> The fix is to pin those dependencies yourself, in both `Move.toml` and `Move.lock`. Replace each

Author: The user should only change Move.toml; a rebuild should update Move.lock I think (we should do an experiment).

Author: [CLAUDE] Fixed — the docs now say to pin in `Move.toml` and rebuild, which updates `Move.lock`. Confirmed by experiment: editing only the manifest's `rev` (A->B) and running `sui move build` re-pinned the lock to B. source-verification.mdx and DESIGN.md.

compare against, the package's original id, and the toolchain version from the package's own
publication metadata (`Published.toml`, or `Move.lock` for packages published before v1.63) for the
environment you are verifying on. Reading the address from the same metadata the package system uses
to link this package means the verifier checks exactly the package a dependent would link against.

Verification runs against whichever network your CLI is pointed at, so switch to the right environment
first:

```bash
sui client switch --env mainnet
sui client verify-source ./my-package
```

On success it prints `Source verification succeeded!` and exits with status 0. Otherwise it lists
every difference it found and exits non-zero.

## What it verifies

Only the **root** package is verified: the bytecode of its modules and its linkage table (the
versions of its dependencies). The command does not check the source of the dependencies themselves.
To verify a dependency, run `verify-source` again against that dependency's own source.

The command performs no compilation itself. It resolves the toolchain version that published the

> package, downloads that release's `sui` binary if it is not already cached under `$MOVE_HOME/binaries`,

Author: it's not $MOVE_HOME anymore

and uses it to build the source. Because that same toolchain produced the onchain bytecode, a matching
build is strong evidence that the source is authentic.

:::info

Verification compares **bytecode**, not source text. Anything that does not affect the compiled output

> (comments, formatting, and the bodies of `macro` functions that no caller invokes) can differ between

Author: ```suggestion
(comments, formatting, and **the bodies of `macro` functions**) can differ between
```

the source and the published package and still verify. The compiler inlines the bodies of called
macros into their callers, so those affect the bytecode and verification covers them. A package that
verifies has bytecode identical to the onchain version.

:::

## Making a package verifiable

> If you publish a package and want others to verify it out of the box:

Author: ```suggestion
If you want to fix a published source package so that it verifies out of the box:
```

> - **Pin every dependency to a commit hash**, not a branch or tag. The current package system does this

Author: ```suggestion
- **Ensure Move.lock has commit hashes**, not branches or tags. The current package system does this
```

  automatically; see [Dependencies are not pinned to commit
  hashes](#dependencies-are-not-pinned-to-commit-hashes) below for older packages.
- **Tag a commit whose source verifies**, so others can check out the precise source. This is not
  necessarily the original publish commit: if you added a `Published.toml` or pinned dependencies to
  make it verify (see below), tag the commit that carries those changes.
- **Record the toolchain version in the publication.** Modern publishing writes it to `Published.toml`.
  If yours does not (see below), you can add it.

If a package's publication does not record its address or toolchain (for example, the package predates
`Published.toml`, or you published it through an offline or multisig flow), you can add a
`Published.toml` next to the manifest that records the onchain address and toolchain version for each
environment. Older `sui` versions ignore this file, so its only effect is to let `verify-source`, and

> anyone depending on the package, find the metadata. Both the verifier and the package system prefer a

Author: ```suggestion
environment. Older `sui` versions ignore this file, so its only effect is to let `verify-source` and
anyone depending on the package find the metadata. Both the verifier and the package system prefer a
```

`Published.toml` over legacy addresses, so the two always agree on the address. It looks like this,
with one section per environment:

```toml
[published.mainnet]
chain-id = "35834a8a"                          # the network's chain id

> published-at = "0x0e735f8c93a95722efd735..."   # the package's current onchain address
> original-id = "0x2c8d603bc51326b8c13cef..."    # its original, first-published address
> version = 8
> toolchain-version = "1.71.1"                   # the sui release that built it
> ```

Author: ````suggestion
toolchain-version = "1.71.1"                   # the sui release that built it

[published.testnet]
...
```
````

```

Use `--all`: publishers frequently publish from a short-lived branch and merge it afterward, so the
commit you want might not be on the main branch at all.

The recording commit is often not the one that was built. A publish records the address *after* the
build, so it can already contain source changes that were never published, and a release tag cut later
can be further ahead again. If a small number of modules still differ there, trust the module list over
the commit: take each named module, walk its history around the publish date, and rebuild with that
file swapped in until the mismatch clears.

```bash
git log --all --oneline <PUBLISH_DATE_RANGE> -- path/to/the/module.move
```

Dedupe the candidates by blob hash (`git rev-parse <COMMIT>:path/to/the/module.move`) so you only
rebuild genuinely distinct sources.

### The toolchain version is not recorded

By default the toolchain version comes from the package's publish metadata. Some publications omit it,
notably packages from before v1.23 and some multisig or offline flows. The command then stops and asks
you to supply the version:

```bash
sui client verify-source <PACKAGE_PATH> --toolchain-version 1.71.1
```

> 

Author: Once you find a version that verifies, you can write it into Published.toml to make future reverification easier.

Author: [CLAUDE] Added after the version-guessing paragraph: "Once you find a version that verifies, record it in the package's `Published.toml` (the `toolchain-version` field) so future verifications need no guessing."

hash used at publication, then rebuild, which updates `Move.lock` to match. For a framework dependency,
[`crates/sui-framework-snapshot/manifest.json`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework-snapshot/manifest.json)
records the revision deployed on the network at publish time, per protocol version; the `sui` release
that was current when the package was published is also a good approximation, because the framework is
backward compatible.

### A dependency does not record a published address

Older toolchains require every dependency to declare its published address, and refuse to build without
it:

```
Failed to publish the Move module(s), reason: Package dependency "MyDep" does not specify a
published address (the Move.toml manifest for "MyDep" does not contain a published-at field).
```

This happens when the pinned revision of a dependency is a plain library checkout whose manifest leaves
its address unset (`my_dep = "_"`), even though the package it links against is onchain. The publisher
usually supplied the address out of band, by depending on a network-specific branch of the dependency
or on a local checkout.

> To rebuild, vendor the dependency: copy it at its pinned revision into your tree, add its onchain

Author: check me on this, but I think you can override the address in the old Move.toml format or use `dep-replacements.<env>.<dep> = { published-at = "...", original-id = "..." }` in the new package system instead of vendoring.

Author: [CLAUDE] Reworked the section to lead with supplying the address via `[dep-replacements.<env>]` (`MyDep = { published-at = "0x...", original-id = "0x..." }`), or the dependency's `published-at` in older manifests, with vendoring kept as the fallback for when you also need source changes. Documented as intended behavior per your note; any bug in applying it is a separate workstream.

address to its manifest, and depend on it by local path.

```toml
[package]
name = "MyDep"
published-at = "0x5306f64e..."   # the dependency's address on this network

[addresses]
my_dep = "0x5306f64e..."
```

Supplying the address yourself does not weaken the check. Verification compares the root package's
linkage table as well as its bytecode, so an address other than the one the package actually links
against causes verification to fail.

### The package cannot be rebuilt from its repository

> Some packages do not build directly from their committed manifest. For example, a build script might

Author: I'd soften this - you'd just have to set up the Move.toml by hand.

source. The command reports which modules differ, and **how many modules differ tells you where to
look**:

- **A few of many modules differ.** The toolchain is almost certainly correct: a mismatched compiler
  perturbs code generation broadly and cannot leave the package's other modules byte-identical. Do not
  sweep toolchain versions. The named modules are the ones whose source drifted, so find the right
  source for those files.
- **Every module differs.** Now the toolchain, the dependencies, or the source tree as a whole is
  suspect.
- **The package has only one module.** This signal is unavailable, so neither outcome tells you
  anything about where the problem lies.

Check out the exact published commit. Because the commit that *records* a package's `published-at` in
source control is often later than the commit the publisher built, search for when the address first
appeared and check out the earliest match:

```bash
git log --all -S <PACKAGE_ID> -- '**/Published.toml' '**/Move.lock'
```

Use `--all`: publishers frequently publish from a short-lived branch and merge it afterward, so the
commit you want might not be on the main branch at all.

> That commit is a starting point, not an answer. A publish records the address *after* the build, so the

Reviewer: nit: reads very AI-esque (specifically the first sentence).

later still can be further ahead again. If a small number of modules continue to differ there, trust
the module list over the commit: take each named module and walk its own history around the publish
date, rebuilding with that file swapped in until the mismatch clears.

```bash
git log --all --oneline <PUBLISH_DATE_RANGE> -- path/to/the/module.move
```

Dedupe the candidates by blob hash (`git rev-parse <COMMIT>:path/to/the/module.move`) so you only
rebuild genuinely distinct sources.

### The toolchain version is not recorded

By default the toolchain version comes from the package's publish metadata. Some publications omit it,
notably packages from before v1.23 and some multisig or offline flows. The command then stops and asks
you to supply the version:

```bash
sui client verify-source <PACKAGE_PATH> --toolchain-version 1.71.1
```

If you do not know the exact version, try the release that was current when the package was published.
Compiler output is stable across neighboring releases, so an adjacent version usually reproduces the
same bytecode; a version from far in the past or future might not, because the bytecode file format
changes occasionally.

> That stability cuts both ways, and it is worth taking seriously before you start guessing. Because a

Reviewer: Also very AI-esque

on a branch that no longer exists, no matter how faithfully its source survives.

To fix this, pin those dependencies in `Move.toml` yourself: replace each branch with the exact commit
hash used at publication, then rebuild, which updates `Move.lock` to match. For a framework dependency,
[`crates/sui-framework-snapshot/manifest.json`](https://github.com/MystenLabs/sui/blob/main/crates/sui-framework-snapshot/manifest.json)
records the revision deployed on the network at publish time, per protocol version; the `sui` release
that was current when the package was published is also a good approximation, because the framework is
backward compatible.

### A dependency does not record a published address

Older toolchains require every dependency to declare its published address, and refuse to build without
it:

```
Failed to publish the Move module(s), reason: Package dependency "Wormhole" does not specify a
published address (the Move.toml manifest for "Wormhole" does not contain a published-at field).
```

This happens when the pinned revision of a dependency is a plain library checkout whose manifest leaves

> its address unset (`wormhole = "_"`), even though the package it links against is onchain. The publisher

Reviewer: switch to referencing a generic, or example package.
